### PR TITLE
fix(bun_shim_impl): pass env to `CreateProcessW`

### PR DESCRIPTION
**Pull request by SunsetTechuila**
### What does this PR do?

fixes https://github.com/oven-sh/bun/issues/11114

fixes https://github.com/oven-sh/bun/issues/11016

fixes https://github.com/oven-sh/bun/issues/9984

fixes https://github.com/oven-sh/bun/issues/9810

fixes https://github.com/oven-sh/bun/issues/9791

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

- [x] I included a test for the new code
- [x] I ran my tests locally and they pass

### CODE REVIEW

File: src/install/windows-shim/bun_shim_impl.zig [hunk 1]
```diff
@@ -795,8 +795,8 @@ fn launcher(comptime mode: LauncherMode, bun_ctx: anytype) mode.RetType() {
             null,
             null,
             1, // true
-            0,
-            null,
+            w.CREATE_UNICODE_ENVIRONMENT,
```
Comment (paperclover) [inline src/install/windows-shim/bun_shim_impl.zig]:
```suggestion
            if (is_standalone) 0 else w.CREATE_UNICODE_ENVIRONMENT,
```

**Comment by SunsetTechuila** [comment]
I guess a new test should be added for this, but idk what it should be

**Comment by Jarred-Sumner** [comment]
@SunsetTechuila a bun install test that installs a local package bin which where all the script does is `process.stdout.write(process.env.npm_config_user_agent)`

File: src/install/windows-shim/bun_shim_impl.zig [hunk 2]
```diff
@@ -795,8 +795,8 @@ fn launcher(comptime mode: LauncherMode, bun_ctx: anytype) mode.RetType() {
...
+            @constCast(bun_ctx.environment),
```
Comment (paperclover) [inline src/install/windows-shim/bun_shim_impl.zig]:
```suggestion
            if (is_standalone) null else @constCast(bun_ctx.environment),
```

In addition, I think `@constCast` might be a subtle mistake.

**Review by paperclover — CHANGES_REQUESTED** [review]
See review comments. This file is built in two modes, so it has to handle one case where a specific env block is provided, but also pass the default (inherit) in standalone.

Otherwise, this is great. Thanks for finding this mistake.

File: src/install/windows-shim/bun_shim_impl.zig (hunk 1 above)
Comment (SunsetTechuila) [reply src/install/windows-shim/bun_shim_impl.zig]: this check confused me, for some reason i thought it was above the code that i changed

https://github.com/oven-sh/bun/blob/fab96a74ea13da04459ea7f62663c4d2fd421778/src/install/windows-shim/bun_shim_impl.zig#L817

**Comment by SunsetTechuila** [comment]
@Jarred-Sumner like this?

**Review by paperclover — APPROVED** [review]

**Comment by Jarred-Sumner** [comment]
Great. I bet there are more issues this fixes.

**Comment by SunsetTechuila** [comment]
> Great. I bet there are more issues this fixes

I did my best to find all

**Comment by Jarred-Sumner** [comment]
the ones related to `Path` or `PATH` i think would be fixed since we're now propagating environment variables correctly

**Comment by SunsetTechuila** [comment]
more issues fixed by this pr: 

- https://github.com/oven-sh/bun/issues/9810
- https://github.com/oven-sh/bun/issues/9791
- https://github.com/oven-sh/bun/issues/10188
- https://github.com/oven-sh/bun/issues/5995

i will try to find as much as i can later and leave a comment with all the relevant issues i found

**Comment by staabm** [comment]
@Jarred-Sumner just a suggestion. at phpstan/phpstan-src we have a so called "issue-bot" which checks after a PR is merged, whether the newly merged stuff changes outcome/output of open issues.

the bot can do such things, because [we have a playground](https://phpstan.org/try) which people can replicate bugs in. they can then store their repro examples and get a sharable unique url for it, e.g.: https://phpstan.org/r/d488550e-4296-4292-8b86-b707b7628044

the issue bot scans open issues which are tagged as "bug" or "feature require" and finds reproduction links in this issues (or comments of this issues). as soon as the outcome/output of a reproduction changes, the issue [bot will leave a comment](https://github.com/phpstan/phpstan/issues/10297#issuecomment-2034488493) with the output diff in the open issue, and tags the initial author of the issue.
that way you can easily see which issues are affected/fixed by merging a PR.

additionally we run this bot in PRs as a separate job, so we can even see before merging which issues will be affected by a given change. 

---

this might be something helpful for your team to get on top of the huge open issue list.

**Comment by SunsetTechuila** [comment]
> i will try to find as much as i can later and leave a comment with all the relevant issues i found

can't find any

**Pull request merged**